assets/src/components/groupedAutocomplete.tsx

```diff
@@ -0,0 +1,731 @@
+        id={id}
+        style={{ userSelect: "none" }}
+      >
+        {heading}
+      </div>
+      <ul role="group" {...props} aria-labelledby={id} />
+    </>
+  )
+}
+
+/**
+ * {@link GroupedAutocomplete} {@link LabelledListboxGroup} component.
+ */
+const GroupOptionList = (props: LabelledListProps) => (
+  <LabelledListboxGroup
+    {...props}
+    headingProps={{ className: "c-autocomplete__group-heading" }}
+    className="c-autocomplete__group-list"
+  />
+)
+// #endregion LabelledList Components
+// #endregion Autocomplete Control Impl
+// #endregion Autocomplete Control
+
+// #region Autocomplete From Search Context
+/**
+ * {@link GroupedAutocompleteFromSearchTextResults} Props
+ */
+interface GroupedAutocompleteFromSearchTextResultsProps
+  extends GroupedAutocompleteControlRefProps,
+    Omit<GroupedAutocompleteProps, "optionGroups"> {
+  /**
+   * Text to search to populate the autocomplete options with.
+   */
+  searchText: string
+  /**
+   * Filters to apply when searching for {@link searchText}.
+   */
+  searchFilters: SearchProperties<boolean>
+  /**
+   * Max number of options to render in a group.
+   */
+  maxElementsPerGroup?: number
+  /**
+   * Callback when a autocomplete vehicle option is selected.
+   * @param selectedOption The selected option vehicle
+   *
+   * ---
+   * @todo
+   *   Potentially rewrite this to be generic across option values.
+   *
+   *   Alternatively provide more callbacks for specific types to avoid
+   *   type deduction.
+   */
+  onSelectVehicleOption: (selectedOption: Vehicle | Ghost) => void
+}
+
+/**
+ * A {@link GroupedAutocomplete} which retrieves it's options from the
+ * {@link searchText} provided and {@link useAutocompleteResults}.
+ */
+export const GroupedAutocompleteFromSearchTextResults = ({
+  onSelectVehicleOption: onSelectVehicleOptionProp,
+  searchText,
+  searchFilters,
+  maxElementsPerGroup = 5,
+  ...props
+}: GroupedAutocompleteFromSearchTextResultsProps) => {
+  const { socket } = useContext(SocketContext)
+  const {
+    vehicle: vehicles,
+    run: runs,
+    operator: operators,
+  } = useAutocompleteResults(
+    socket,
+    searchText,
+    searchFilters,
+    maxElementsPerGroup
+  )
+
+  const onSelectVehicleOption = (selectedOption: Vehicle | Ghost) => () => {
+    onSelectVehicleOptionProp(selectedOption)
+  }
+
+  // Build groups and options from search results.
+  const groups = [
+    {
+      group: {
+        title: <h2>{searchPropertyDisplayConfig.vehicle.name}</h2>,
+        options: vehicles
+          .slice(0, maxElementsPerGroup)
+          .map((v) =>
+            autocompleteOptionData(
+              (isVehicle(v) && v.label) || v.id,
+              onSelectVehicleOption(v)
+            )
+          ),
+      },
+    },
+    {
+      group: {
+        title: <h2>{searchPropertyDisplayConfig.operator.name}</h2>,
+        options: operators
+          .slice(0, maxElementsPerGroup)
+          .filter(isVehicle)
```

Comment:
question: Should this filtering happen before slicing?

Reply:
a5b9495a6bf2bc0910e4a1633fbd70fce141f283